src/mapping/VolumeCellInterpolation.cpp

```diff
@@ -0,0 +1,114 @@
+#include "mapping/Mapping.hpp"
+#include "mapping/Polation.hpp"
+#include "math/differences.hpp"
+#include "mesh/Data.hpp"
+#include "mesh/Mesh.hpp"
+#include "mesh/SharedPointer.hpp"
+#include "mesh/Vertex.hpp"
+#include "query/Index.hpp"
+#include "utils/Event.hpp"
+#include "utils/Statistics.hpp"
+#include "utils/assertion.hpp"
+
+namespace precice {
+extern bool syncMode;
+
+namespace mapping {
+
+VolumeCellInterpolation::VolumeCellInterpolation(
+    Constraint constraint,
+    int        dimensions)
+    : BarycentricBaseMapping(constraint, dimensions)
+{
+  if (constraint == CONSISTENT) {
+    setInputRequirement(Mapping::MeshRequirement::FULL);
+    setOutputRequirement(Mapping::MeshRequirement::VERTEX);
+  } else if (constraint == CONSERVATIVE) {
+    setInputRequirement(Mapping::MeshRequirement::VERTEX);
+    setOutputRequirement(Mapping::MeshRequirement::FULL);
+  } else {
+    PRECICE_ASSERT(constraint == SCALEDCONSISTENT, constraint);
+    setInputRequirement(Mapping::MeshRequirement::FULL);
+    setOutputRequirement(Mapping::MeshRequirement::FULL);
+  }
+
+  PRECICE_ASSERT(constraint != SCALEDCONSISTENT, "Volume mapping doesn't support scaled-consistent mappings.");
+}
+
+void VolumeCellInterpolation::computeMapping()
+{
+  PRECICE_TRACE(input()->vertices().size(), output()->vertices().size());
+  const std::string     baseEvent = "map.vci.computeMapping.From" + input()->getName() + "To" + output()->getName();
+  precice::utils::Event e(baseEvent, precice::syncMode);
+  PRECICE_ASSERT(getDimensions() == 2, "Volume mapping not available in 3D.");
+
+  // Setup Direction of Mapping
+  mesh::PtrMesh origins, searchSpace;
+  if (hasConstraint(CONSERVATIVE)) {
+    PRECICE_DEBUG("Compute conservative mapping");
+    origins     = input();
+    searchSpace = output();
+  } else {
+    PRECICE_DEBUG("Compute consistent mapping");
+    origins     = output();
+    searchSpace = input();
+  }
+
+  const auto &fVertices = origins->vertices();
+
+  if (getDimensions() == 2) {
+    if (!fVertices.empty() && searchSpace->triangles().empty()) {
+      PRECICE_WARN("2D Mesh \"{}\" does not contain triangles. "
+                   "Volume cell interpolation falls back to nearest neighbor mapping.",
+                   searchSpace->getName());
+    }
+  } else {
+    if (!fVertices.empty() && searchSpace->triangles().empty()) {
+      PRECICE_WARN("3D Mesh \"{}\" does not contain tetrahedra. "
+                   "Volume cell interpolation will map to primitives of lower dimension.",
+                   searchSpace->getName());
+    }
+  }
+
+  // Amount of nearest elements to fetch for detailed comparison.
+  // This safety margin results in a candidate set which forms the base for the
+  // local nearest projection and counters the loss of detail due to bounding box generation.
+  // @TODO Add a configuration option for this factor
+  constexpr int nnearest = 4;
+
+  auto                                  &index = searchSpace->index();
```

Comment:
```suggestion
 const auto                                  &index = searchSpace->index();
```
?

Reply:
Change to do if we ever refactor to make all the sub-calls const too